src/core/storage/sframe_data/join_impl.cpp

```diff
@@ -156,67 +161,188 @@ hash_join_executor::hash_join_executor(const sframe &left,
     _right_join = true;
   }
 
+  ASSERT_EQ(_left_join_positions.size(), _right_join_positions.size());
+
+  /* handy script for reverse lookup */
+  for(size_t i = 0; i < _left_join_positions.size(); ++i) {
+    auto ret = _right_to_left_join_positions.emplace(_right_join_positions[i],
+                                                     _left_join_positions[i]);
+    ASSERT_TRUE(ret.second);
+
+    ret = _left_to_right_join_positions.emplace(_left_join_positions[i],
+                                                _right_join_positions[i]);
+    ASSERT_TRUE(ret.second);
+  }
+
+  /*
+   * using names from sframe: unique and non-empty
+   * the new names in result sframe should have no confilict
+   *
+   * this is order dependent; doing in reverse will result in
+   * different result.
+   *
+   * since _alter_name_right only applies to original order,
+   * we have to obtain the original construction
+   **/
+
+  // get orginal left sframe's column names
+  auto col_names_original_order = left.column_names();
+  _output_column_names = left.column_names();
+  _output_column_types = left.column_types();
+
+  std::set<std::string> new_table = {
+      std::make_move_iterator(col_names_original_order.begin()),
+      std::make_move_iterator(col_names_original_order.end())};
+
+  // get orginal right sframe's column names
+  col_names_original_order = right.column_names();
+
+  /* check the name resolution is order dependent or not.
+   * resolution {A -> B} is order dependent iff
+   * 1. A belongs to right.column_names()
+   * 2. B is not equal to keys from right.column_names()
+   * 3. B should be unique to other B'
+  */
+  std::set<std::string> unique_names_from_right = {
+      std::begin(col_names_original_order), std::end(col_names_original_order)};
+
+  std::set<std::string> unique_names_from_user;
+  for (auto& entry : _alter_names_right) {
+    if (unique_names_from_right.count(entry.first) == 0) {
+          std::stringstream ss;
+          ss << "user provided column name { " << entry.first
+             << " } is not found in right SFrame.";
+          log_and_throw(ss.str());
+    }
+
+    if (!unique_names_from_user.insert(entry.second).second) {
+          std::stringstream ss;
+          ss << "user provided resolution name { " << entry.second
+             << " } duplicates with other resolution name.";
+          log_and_throw(ss.str());
+    }
+
+    if (unique_names_from_right.count(entry.second)) {
+          std::stringstream ss;
+          ss << "user provided resolution name { " << entry.second
+             << " } is not allowed to be same with any name in right SFrame.";
+          log_and_throw(ss.str());
+    }
+  }
+
+  // check the original right sframe columns
+  for (auto& name : col_names_original_order) {
```

Comment:
Minor: Let's use const auto& where appropriate -- use auto& only if you explicitly want to change the contents present.

Reply:
Got it. For this one, name is actually used and will be moved in some conditions.